redbot/core/modlog.py

```diff
@@ -51,12 +57,94 @@
     _conf.register_custom(_CASETYPES)
     _conf.register_custom(_CASES)
     await _migrate_config(from_version=await _conf.schema_version(), to_version=_SCHEMA_VERSION)
+    await register_casetypes(all_generics)
+
+    async def on_member_ban(guild: discord.Guild, member: discord.Member):
+
+        if not guild.me.guild_permissions.view_audit_log:
+            return
+
+        try:
+            await get_modlog_channel(guild)
+        except RuntimeError:
+            return  # No modlog channel so no point in continuing
+
+        when = datetime.utcnow()
+        before = when + timedelta(minutes=1)
+        after = when - timedelta(minutes=1)
+
+        mod, reason, date = None, None, None
+        attempts = 0
+
+        while mod is None and attempts < 12:  # wait up to an hour to find a matching case
+            attempts += 1
+            try:
+                entry = await guild.audit_logs(
+                    action=discord.AuditLogAction.ban, before=before, after=after
+                ).find(lambda e: e.target.id == member.id)
+            except discord.Forbidden:
+                break
+            except discord.HTTPException:
+                pass
+            else:
+                if entry:
+                    if entry.user.id == guild.me.id:
+                        # Don't create modlog entires for the bot's own bans, cogs do this.
+                        return
+                    mod, reason, date = entry.user, entry.reason, entry.created_at
+
+            await asyncio.sleep(300)
+
+        if mod:
+            await create_case(_bot_ref, guild, date, "ban", member, mod, reason)
+
+    async def on_member_unban(guild: discord.Guild, user: discord.User):
+        if not guild.me.guild_permissions.view_audit_log:
+            return
+
+        try:
+            await get_modlog_channel(guild)
+        except RuntimeError:
+            return  # No modlog channel so no point in continuing
+
+        when = datetime.utcnow()
+        before = when + timedelta(minutes=1)
+        after = when - timedelta(minutes=1)
+
+        mod, reason, date = None, None, None
+        attempts = 0
+
+        while mod is None and attempts < 12:  # wait up to an hour to find a matching case
+            attempts += 1
+            try:
+                entry = await guild.audit_logs(
+                    action=discord.AuditLogAction.unban, before=before, after=after
+                ).find(lambda e: e.target.id == user.id)
+            except discord.Forbidden:
+                break
+            except discord.HTTPException:
+                pass
+            else:
+                if entry:
+                    if entry.user.id == guild.me.id:
+                        # Don't create modlog entires for the bot's own unbans, cogs do this.
+                        return
+                    mod, reason, date = entry.user, entry.reason, entry.created_at
```

Comment:
needs a break after this (and other) to prevent sleeping again before making the case

Reply:
I went for a slight restructure of the loop instead.